pd/src/app.rs

```diff
@@ -160,33 +175,192 @@ impl App {
+    ///
+    /// In the transaction validation performed before adding a transaction into the
+    /// mempool, we check that:
+    ///
+    /// * All binding and auth sigs signatures verify (stateless),
+    /// * All proofs verify (stateless and stateful),
+    /// * The transaction does not reveal nullifiers already revealed in another transaction
+    /// in the mempool or in the database,
+    ///
+    /// If a transaction does not pass these checks, we return a non-zero `CheckTx` response
+    /// code, and the transaction will not be added into the mempool.
+    ///
+    /// We do not queue up any state changes into `PendingBlock` until `DeliverTx` where these
+    /// checks are repeated.
     #[instrument(skip(self))]
-    fn deliver_tx(&mut self, txbytes: Bytes) -> response::DeliverTx {
-        // TODO: implement (#135)
-
-        // Transactions that cannot be deserialized should return a non-zero `DeliverTx` code.
-        let _transaction = match Transaction::try_from(txbytes.as_ref()) {
-            Ok(transaction) => transaction,
-            Err(_) => {
-                return response::DeliverTx {
-                    code: 1,
-                    ..Default::default()
+    fn check_tx(
+        &mut self,
+        request: request::CheckTx,
+    ) -> impl Future<Output = Result<Response, BoxError>> {
+        let state = self.state.clone();
+        let mempool_nullifiers = self.mempool_nullifiers.clone();
+        let nct_root = self.note_commitment_tree.root2();
+        let finished_signal = self.completion_tracker.start();
+
+        async move {
+            let transaction = match Transaction::try_from(request.tx.as_ref()) {
+                Ok(transaction) => transaction,
+                Err(_) => {
+                    return Ok(Response::CheckTx(response::CheckTx {
+                        code: 1,
+                        ..Default::default()
+                    }))
+                }
+            };
+
+            let pending_transaction = match transaction.verify_stateless(None) {
+                Ok(pending_transaction) => pending_transaction,
+                Err(_) => {
+                    return Ok(Response::CheckTx(response::CheckTx {
+                        code: 1,
+                        ..Default::default()
+                    }))
+                }
+            };
+
+            // Ensure we do not add any transactions with duplicate nullifiers into the mempool.
+            //
+            // Note that we only run this logic if this `CheckTx` request is from a new transaction
+            // (i.e. `CheckTxKind::New`). If this is a recheck of an existing entry in the mempool,
+            // then we don't need to add the nullifier again, as it's already in `self.mempool_nullifiers`.
+            // Rechecks occur whenever a block is committed if the Tendermint `mempool.recheck` option is
+            // true, which is the default option.
+            if request.kind == CheckTxKind::New {
```

Comment:
Another option here would be to additional store the transaction ID in addition to the nullifier in the `mempool_nullifiers` structure so we can verify that if there is a matching nullifier already in the mempool, it's from the same transaction ID